### ISSUE TITLE
Idea for data bag: chaining of validators and transformers

### ISSUE THREAD

**Issue by ebuildy** [symfony/symfony#22674]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 4.0.0

HTTP is ``request -> APP -> output``, form is very good to handle complex request, but (I found) there is no simple way to deal with simple request. Thats why I have built a very little class that can use Symfony validators and transformers, in a chain fashion, to transform HTTP request fields to "business" data.

Example: a ``recoverPasswordAction`` method, that take an email, check user exists and send recover email, here the syntax I proprose:

```php
   /**
     * @Route("/forgot-password")
     */
    public function recoverPasswordAction(Request $request)
    {
        try {
        /** @var \...\Entity\User $user */
        $user = (new SupplyDataChain($this->get('validator')))
                        ->setValue($request, 'email')
                        ->validate(new NotBlank()) // email field is here
                        ->validate(new Email()) // email field is a valid email
                        ->transform([$this->userModel, 'findOneByEmail']) // Get user from Doctrine repo.
                        ->validate(new NotBlank( ['message' => 'Account not found!'] )) // Use is here ?
                        ->getValue(); // Get value
        .....
        } catch(\Exception $e) { ......
...
}
```

(Here a simple implementation https://gist.github.com/ebuildy/32416d5bbd5b26ef82b15849c9616716, I am using it in several project and lot of friend found this very helpful, thats why I am sharing it here).

**Comment by ro0NL** [comment]
form is also good for simple request :))

validating, transforming.. sounds like a form usecase to me. Note the term 'form' is a bit misleading, as it's practically just request validation, i.e. your proposal.

**Comment by ebuildy** [comment]
Yeah this is possible with form, but I found the syntax very long to write and I believe next Symfony will be "RAD" ^^, form is totally opposite (I never found an example where multiple validators and transformers, 100% in-line without tons of Php files).

in your example, you don't check $user exists, I guess you must write another line:

```
if (empty($user)) { throw new \Exception(....) }
```

whereas chaining keep it clear, no ``variables`` created, cascade of ``if`` blocks etc...

**Comment by ro0NL** [comment]
>  in your example, you don't check $user exists

Well.. it depends if user is optional yes/no. If you validate a model containing the user, it would still be just a constraint. Otherwise, if you expect it, the throw happens where the request is mapped to a model.

> next Symfony will be "RAD" ^^, form is totally opposite

This is your perspective really, we do RAD with symfony forms... Yes it's frustrating sometimes (when it's get complex actually, and mostly is frontend related), but it's a scalable approach (which is important for RAD, if the project succeeds).

I'd rather refactor at form level than chaining the total approach (from tons of datachains to _a_ form).

**Comment by flip111** [comment]
@ebuildy I like your idea for the code. You say that ro0NL doesn't check if the user exist, but how i read your code you also don't check if the user exist. After you call this `->transform([$this->userModel, 'findOneByEmail'])` it should turn an email into a user. But even when no user was returned you pass it to the validator `->validate(new NotBlank( ['message' => 'Account not found!'] ))`. So the "don't have a user check" is now done by the validator i guess? I would prefer to check this explicit and not depend on the implementation of the validator itself.
